### ISSUE TITLE
03. PR3: raw/view projection と Gaussian consistency target helper を追加する

### ISSUE THREAD

**Issue by ozw4** [ozw4/seisai#434]
# 03. PR3: raw/view projection と Gaussian consistency target helper を追加する

## 目的

memory bank に保存する raw pick index と、現在 batch の view coordinate を相互変換する helper を追加する。さらに、memory pick から time-axis Gaussian target を作る helper を追加する。

この issue では helper のみを実装する。criterion 本体は後続 issue で実装する。

## 依存 issue

- `01_issue_add_trace_memory_module_contract.md`
- `02_issue_implement_sparse_trace_pick_memory_bank.md`

## 対象ファイル

```text
packages/seisai-engine/src/seisai_engine/pipelines/fbpick/coarse/trace_memory_consistency.py
```

## 背景

memory bank には view bin ではなく **raw sample index** を保存する。primary key、secondary key、time crop、time stretch、trace_tshift が変わっても同じ raw trace の pick を比較できるようにするため。

current view で loss を作るときは、memory の raw pick を current view に投影する。

```text
memory_view = memory_raw * factor - start + trace_tshift_view
```

current prediction を memory に保存するときは逆変換する。

```text
current_raw = (current_view_pick + start - trace_tshift_view) / factor
```

### 1. meta tensor 取得 helper

```python
def get_trace_memory_projection_meta(
    batch: dict,
    *,
    device: torch.device,
    B: int,
    H: int,
) -> tuple[torch.Tensor, torch.Tensor, torch.Tensor, torch.Tensor]:
    ...
```

戻り値:

```text
factor:             (B,)   float32
start:              (B,)   float32
trace_tshift_view:  (B,H)  float32
trace_valid:        (B,H)  bool
```

入力元:

```text
batch['meta']['factor']
batch['meta']['start']
batch['meta']['trace_tshift_view']  # optional。なければ zeros(B,H)
batch['meta']['trace_valid']
```

validation:

```text
factor は finite かつ > 0
start は finite
trace_tshift_view は finite
shape mismatch は ValueError
```

### 2. raw -> view projection

```python
def project_memory_raw_pick_to_view(
    pick_raw: torch.Tensor,
    *,
    sample_index: torch.Tensor,
    row_index: torch.Tensor,
    factor: torch.Tensor,
    start: torch.Tensor,
    trace_tshift_view: torch.Tensor,
) -> torch.Tensor:
    ...
```

shape:

```text
pick_raw:     (N,)
sample_index: (N,)  # batch index b
row_index:    (N,)  # trace row h
return:       (N,)
```

式:

```python
pick_view = pick_raw * factor[sample_index] - start[sample_index] + trace_tshift_view[sample_index, row_index]
```

### 3. view -> raw projection

```python
def project_current_view_pick_to_raw(
    pick_view: torch.Tensor,
    *,
    sample_index: torch.Tensor,
    row_index: torch.Tensor,
    factor: torch.Tensor,
    start: torch.Tensor,
    trace_tshift_view: torch.Tensor,
) -> torch.Tensor:
    ...
```

式:

```python
pick_raw = (pick_view + start[sample_index] - trace_tshift_view[sample_index, row_index]) / factor[sample_index]
```

round-trip は float 誤差を除いて成立する。

### 4. sigma_ms -> sigma_bins

```python
def trace_memory_sigma_bins(
    *,
    sigma_ms: float,
    batch: dict,
    sample_index: torch.Tensor,
    device: torch.device,
) -> torch.Tensor:
    ...
```

`batch['meta']['dt_eff_sec']` を使う。shape は `(B,)` または scalar を許す。

```text
sigma_bins = sigma_ms / (dt_eff_sec[sample_index] * 1000.0)
```

validation:

```text
sigma_ms > 0
meta.dt_eff_sec が finite かつ > 0
sigma_bins は finite かつ > 0
```

数値安定化として `sigma_bins.clamp_min(1.0e-6)` は許可する。

### 5. Gaussian target

```python
def build_trace_memory_gaussian_targets(
    *,
    centers: torch.Tensor,
    sigma_bins: torch.Tensor,
    W: int,
) -> torch.Tensor:
    ...
```

shape:

```text
centers:    (N,)
sigma_bins: (N,) or scalar
return:     (N,W)
```

実装:

```python
grid = torch.arange(W, device=centers.device, dtype=torch.float32)
q = torch.exp(-0.5 * ((grid[None, :] - centers[:, None]) / sigma_bins[:, None]) ** 2)
q = q / q.sum(dim=-1, keepdim=True).clamp_min(1.0e-12)
```

center は整数に丸めない。float center のまま Gaussian を作る。

`centers` が view 範囲外の row はこの helper の呼び出し前に mask する。helper 内では `0 <= center <= W-1` を validation してよい。

## 非目標

この issue では以下を実装しない。

- memory lookup/update
- current logits との cross entropy/KL 計算
- label_guard
- pmax / min_count / age gate
- criterion wrapper
- train.py 接続

## 受け入れ条件

- `raw -> view -> raw` の round-trip test が通る。
- `trace_tshift_view` がある場合とない場合の両方で projection が動く。
- `dt_eff_sec` から sigma_bins を正しく計算できる。
- Gaussian target は各 row で sum=1 になる。
- center が float の場合でも target の peak が中心付近になる。

## 確認コマンド

```bash
python -m py_compile \
  packages/seisai-engine/src/seisai_engine/pipelines/fbpick/coarse/trace_memory_consistency.py
```
